lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,832 @@
+  let closed = false;
+
+  const writable = new Writable({
+    highWaterMark,
+    objectMode,
+    decodeStrings,
+    signal,
+
+    writev(chunks, callback) {
+      function done(error) {
+        try {
+          callback(error);
+        } catch (error) {
+          process.nextTick(() => writable.destroy(error));
+        }
+      }
+
+      PromisePrototypeThen(
+        writer.ready,
+        () => {
+          return PromisePrototypeThen(
+            PromiseAll(
+              ArrayPrototypeMap(
+                chunks,
+                (chunk) => writer.write(chunk))),
+            done,
+            done);
+        },
+        done);
+    },
+
+    write(chunk, encoding, callback) {
+      if (typeof chunk === 'string' && decodeStrings && !objectMode)
+        chunk = new Uint8Array(Buffer.from(chunk, encoding));
+
+      function done(error) {
+        try {
+          callback(error);
+        } catch (error) {
+          writable.destroy(error);
+        }
+      }
+
+      PromisePrototypeThen(
+        writer.ready,
+        () => {
+          return PromisePrototypeThen(
+            writer.write(chunk),
+            done,
+            done);
+        },
+        done);
+    },
+
+    destroy(error, callback) {
+      function done() {
+        try {
+          callback(error);
+        } catch (error) {
+          process.nextTick(() => { throw error; });
+        }
+      }
+
+      if (!closed) {
+        if (error != null) {
+          PromisePrototypeThen(
+            writer.abort(error),
+            done,
+            done);
+        } else {
+          PromisePrototypeThen(
+            writer.close(),
+            done,
+            done);
+        }
+        return;
+      }
+
+      done();
+    },
+
+    final(callback) {
+      function done(error) {
+        try {
+          callback(error);
+        } catch (error) {
+          process.nextTick(() => writable.destroy(error));
+        }
+      }
+
+      if (!closed) {
+        PromisePrototypeThen(
+          writer.close(),
+          done,
+          done);
+      }
+    },
+  });
+
+  PromisePrototypeThen(
+    writer.closed,
+    () => {
+      // If the WritableStream closes before the stream.Writable has been
+      // ended, we signal an error on the stream.Writable.
+      closed = true;
+      if (!writable.writableEnded)
+        writable.destroy(new ERR_STREAM_PREMATURE_CLOSE());
+    },
+    (error) => {
+      // If the WritableStream errors before the stream.Writable has been
+      // destroyed, signal an error on the stream.Writable.
+      closed = true;
+      writable.destroy(error);
+    });
+
+  return writable;
+}
+
+/**
+ * @param {Readable} streamReadable
+ * @returns {ReadableStream}
+ */
+function newReadableStreamFromStreamReadable(streamReadable) {
+  if (typeof streamReadable?._readableState !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'streamReadable',
+      'stream.Readable',
+      streamReadable);
+  }
```

Comment:
Throw is `streamReadable.didRead`? https://github.com/nodejs/node/pull/36820 How does web streams handle this if it hasn't been teed? Should we apply same semantics here?

Reply:
Once that lands, yes, we can add it. Once `getReader()` is called on a `ReadableStream` it is `locked`, and all other attempts to consume it will fail. We can do that here also. If `streamreadable.readableDidRead` is true when the adapter is being created, the call will throw.

Reply:
I'm thinking `stream.isDestroyed` should be the same behavior as having been locked/consumed. Not sure if we should apply something similar to writable?